[WK2] Further simplify ArgumentCoder decoding for std::tuple<>, std::array<>

zdobersek opened the pull request:
#### 22987c3dcd7be882ad32ee1a0920e35137a6eadc
<pre>
[WK2] Further simplify ArgumentCoder decoding for std::tuple&lt;&gt;, std::array&lt;&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=247227">https://bugs.webkit.org/show_bug.cgi?id=247227</a>

Reviewed by Kimmo Kinnunen.

Have the std::tuple&lt;&gt; decoding in ArgumentCoders.h avoid constructing tuples of
forwarding references by instead passing the rvalue references to the decoded
elements through a parameter pack. Recursion is still used, and the rvalue
references are passed on until the size of the parameter pack matches the size
of the tuple, at which point said tuple is constructed and returned as the result
of the decoding.

Identical approach is adopted for the std::array&lt;&gt; decoder.

The change incorporates a MSCV++ workaround for a compiler bug which breaks
builds due to the constexpr if-statement checking the current index against the
number of the already-decoded elements. Workaround moves the result of this check
into a separate constexpr boolean. Tracking is set up in bug #247226.

* Source/WebKit/Platform/IPC/ArgumentCoders.h:

Canonical link: <a href="https://commits.webkit.org/256231@main">https://commits.webkit.org/256231@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bb066c16c44e7460044fb93500db97037cfc5a95

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95154 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4299 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/28170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104752 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4395 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33158 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87472 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100648 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100822 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3200 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81700 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30173 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/86941 "Build was cancelled. Recent messages:") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73071 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38884 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18498 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36703 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19779 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4306 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40639 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42480 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42622 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39020 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/Platform/IPC/ArgumentCoders.h:
```diff
@@ -319,26 +319,21 @@ template<typename... Elements> struct ArgumentCoder<std::tuple<Elements...>> {
     {
-        constexpr size_t Index = sizeof...(Indices);
-        static_assert(Index == std::tuple_size_v<OptionalTuple>);
+        constexpr size_t Index = sizeof...(DecodedTypes);
```
I know I missed it the last time, but now that you're touching the lines:
Maybe the constexpr constants could follow normal webkit variable naming. E.g. `index`, `shouldHandleElement`, etc.
shouldHandleElement has extra parenthesis.

zdobersek replied:
OK.

kkinnunen-apple reviewed (approved):
Makes sense but I cannot say I could have written it myself

zdobersek committed:
[WK2] Further simplify ArgumentCoder decoding for std::tuple<>, std::array<>
https://bugs.webkit.org/show_bug.cgi?id=247227

Reviewed by Kimmo Kinnunen.

Have the std::tuple<> decoding in ArgumentCoders.h avoid constructing tuples of
forwarding references by instead passing the rvalue references to the decoded
elements through a parameter pack. Recursion is still used, and the rvalue
references are passed on until the size of the parameter pack matches the size
of the tuple, at which point said tuple is constructed and returned as the result
of the decoding.

Identical approach is adopted for the std::array<> decoder.

The change incorporates a MSCV++ workaround for a compiler bug which breaks
builds due to the constexpr if-statement checking the current index against the
number of the already-decoded elements. Workaround moves the result of this check
into a separate constexpr boolean. Tracking is set up in bug #247226.

* Source/WebKit/Platform/IPC/ArgumentCoders.h:

Canonical link: https://commits.webkit.org/256231@main
